packages/block-library/src/image/edit.js

```diff
@@ -188,6 +188,9 @@ class ImageEdit extends Component {
 			width: undefined,
 			height: undefined,
 		} );
+
+		// Update target link in case the default has been changed (e.g. by a plugin).
+		this.onSetLinkDestination( this.props.attributes.linkDestination );
```

Comment:
This results in two consecutive calls to `setAttributes`, which is not great because every call to `setAttributes` creates a new undo state for the user. Ideally all attributes updates would happen in the call to setAttributes on line 186. You'd probably need to extract the contents of `onSetLinkDestination` into a separate function that returns an object that can be merged into the state (the same way that `pickRelevantMediaFiles` works).

Reply:
Thanks. I've looked into this and it's not as simple as it seems. I split the functions as you said...

```php
	onSelectImage( media ) {
		if ( ! media || ! media.url ) {
			this.props.setAttributes( {
				url: undefined,
				alt: undefined,
				id: undefined,
				caption: undefined,
			} );
			return;
		}

		this.setState( {
			isEditing: false,
		} );

		this.props.setAttributes( {
			...pickRelevantMediaFiles( media ),
			...this.buildLinkAttributes( this.props.attributes.linkDestination ),
			width: undefined,
			height: undefined,
		} );
	}

	onSetLinkDestination( value ) {
		this.props.setAttributes( this.buildLinkAttributes( value ) );
	}

	buildLinkAttributes( destination ) {
		let href;

		if ( destination === LINK_DESTINATION_NONE ) {
			href = undefined;
		} else if ( destination === LINK_DESTINATION_MEDIA ) {
			href = ( this.props.image && this.props.image.source_url ) || this.props.attributes.url;
		} else if ( destination === LINK_DESTINATION_ATTACHMENT ) {
			href = this.props.image && this.props.image.link;
		} else {
			href = this.props.attributes.href;
		}

		return {
			linkDestination: destination,
			href,
		};
	}
```

...but it didn't work - the URL field was empty after selecting an image. I think the problem is that the callbacks bound to the attributes set by `onSelectImage` (using keys from `pickRelevantMediaFiles`) change the values used later by my explicit call to `onSetLinkDestination`. By combining the attribute updates from `pickRelevantMediaFiles` and `onSetLinkDestination` into one `setAttributes` call, like in the above code, the values aren't the same as in the original PR because some haven't yet updated by that part of the execution. Splitting the logic across the two calls to `setAttributes` accidentally worked pretty well to avoid this.

One way to fix this is to explicitly pass the relevant parameters into my new function, e.g. pass `media` in to `buildLinkAttributes`, but this is getting a bit messy. Is there instead some way to set attributes, letting all of the callbacks bound to those attribute fire, before "committing" the attributes later to generate the undo action? Or do you have any other ideas for solving this?

Reply:
Ah, that's unfortunate. 

> Is there instead some way to set attributes, letting all of the callbacks bound to those attribute fire, before "committing" the attributes later to generate the undo action?

Kind of like a batch update. As far as I'm aware nothing like that exists.

> One way to fix this is to explicitly pass the relevant parameters into my new function

That seems like the best option, maybe it just accepts the full attributes. Something like this could be the usage:
```
const mediaAttributes = pickRelevantMediaFiles( media );
const linkAttributes = buildLinkAttributes( 
    this.props.attributes.linkDestination,
    { ...this.props.attributes, ...mediaAttributes }
);
this.props.setAttributes( {
    ...mediaAttributes,
    ...linkAttributes,
    width: undefined,
    height: undefined,
} );
```

Usage in `onSetLinkDestination` would still be fairly similar:
```
this.props.setAttributes( this.buildLinkAttributes( value, this.props.attributes ) );
```